misc-scripts/xref_mapping/XrefParser/RefSeqGPFFParser.pm

```diff
@@ -17,340 +17,341 @@ limitations under the License.
 
 =cut
 
-# Parse RefSeq GPFF files to create xrefs.
-
 package XrefParser::RefSeqGPFFParser;
 
 use strict;
 use warnings;
 use Carp;
-use File::Basename;
-
-use base qw( XrefParser::BaseParser );
-my $peptide_source_id;
-my $mrna_source_id ;
-my $ncrna_source_id ;
-my $pred_peptide_source_id ;
-my $pred_mrna_source_id ;
-my $pred_ncrna_source_id ;
-my $entrez_source_id;
-my $wiki_source_id;
-my %entrez;
+use List::MoreUtils qw(uniq);
+use Readonly;
 
-sub run {
+use parent qw( XrefParser::BaseParser );
+
+# Refseq sources to consider. Prefixes not in this list will be ignored
+Readonly my $REFSEQ_SOURCES => {
+    NM => 'RefSeq_mRNA',
+    NR => 'RefSeq_ncRNA',
+    XM => 'RefSeq_mRNA_predicted',
+    XR => 'RefSeq_ncRNA_predicted',
+    NP => 'RefSeq_peptide',
+    XP => 'RefSeq_peptide_predicted',
+};
 
+
+sub run {
   my ($self, $ref_arg) = @_;
+
   my $source_id    = $ref_arg->{source_id};
   my $species_id   = $ref_arg->{species_id};
   my $species_name = $ref_arg->{species};
   my $files        = $ref_arg->{files};
   my $release_file = $ref_arg->{rel_file};
-  my $verbose      = $ref_arg->{verbose};
-  my $dbi          = $ref_arg->{dbi};
-  $dbi = $self->dbi unless defined $dbi;
+  my $dbi          = $ref_arg->{dbi} // $self->dbi;
+  my $verbose      = $ref_arg->{verbose} // 0;
 
   if((!defined $source_id) or (!defined $species_id) or (!defined $files)){
     croak "Need to pass source_id, species_id, files and rel_file as pairs";
   }
-  $verbose |=0;
-
-  my @files = @{$files};
-
-
-  $peptide_source_id =
-    $self->get_source_id_for_source_name('RefSeq_peptide', undef, $dbi);
-  $mrna_source_id =
-    $self->get_source_id_for_source_name('RefSeq_mRNA','refseq', $dbi);
-  $ncrna_source_id =
-    $self->get_source_id_for_source_name('RefSeq_ncRNA', undef, $dbi);
-
-  $pred_peptide_source_id =
-    $self->get_source_id_for_source_name('RefSeq_peptide_predicted', undef, $dbi);
-  $pred_mrna_source_id =
-    $self->get_source_id_for_source_name('RefSeq_mRNA_predicted','refseq', $dbi);
-  $pred_ncrna_source_id =
-    $self->get_source_id_for_source_name('RefSeq_ncRNA_predicted', undef, $dbi);
-
-  $entrez_source_id = $self->get_source_id_for_source_name('EntrezGene', undef, $dbi);
-  $wiki_source_id = $self->get_source_id_for_source_name('WikiGene', undef, $dbi);
-
-  if($verbose){
-    print "RefSeq_peptide source ID = $peptide_source_id\n";
-    print "RefSeq_mRNA source ID = $mrna_source_id\n";
-    print "RefSeq_ncRNA source ID = $ncrna_source_id\n";
-    print "RefSeq_peptide_predicted source ID = $pred_peptide_source_id\n";
-    print "RefSeq_mRNA_predicted source ID = $pred_mrna_source_id\n" ;
-    print "RefSeq_ncRNA_predicted source ID = $pred_ncrna_source_id\n" ;
+
+  $self->{species_id} = $species_id;
+  $self->{dbi} = $dbi;
+  $self->{verbose} = $verbose;
+
+  # get RefSeq source ids
+  while (my ($source_prefix, $source_name) = each %{$REFSEQ_SOURCES}) {
```

Comment:
I like that these things are all packed together instead of being in 10 different arguments. Using the self hash to store it all is fine but you could also package them into an object, say `$reference_info`, and pass it to `xref_from_record` this way. You'll keep the benefits, it will be clear that `xref_from_record` needs all that reference info. If you also pass `$dbi` as argument then `xref_from_record` will become a pure function if you're into that.